File: src/cmcandy/solutions/_0002.java
package cmcandy.solutions;

/**
 * Created by CHEN on 2020/4/4.
 */
public class _0002 {
    /**
     给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。

     如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。

     您可以假设除了数字 0 之外，这两个数都不会以 0 开头。

     示例：

     输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
     输出：7 -> 0 -> 8
     原因：342 + 465 = 807

     */

    /**
     * Definition for singly-linked list.
     **/
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    class Solution {
        public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
            ListNode answer = l1;
            int add = 0;
            //如果l1.next ==null则，l1
            while (l2 != null && l1 != null) {
                l1.val += l2.val;

                if (add == 1) {
                    l1.val += 1;
                    add = 0;
                }
                if (l1.val >= 10) {
                    l1.val -= 10;
                    add = 1;
                }
                if (l1.next == null && add == 1) {
                    l1.next = new ListNode(1);
                    add = 0;
                } else if (l1.next == null && l2.next != null) {
                    l1.next = new ListNode(0);
                } else if (l2.next == null && add == 1) {
                    l2.next = new ListNode(0);
                }
                l1 = l1.next;
                l2 = l2.next;
            }
            return answer;
        }
    }


}

File: src/cmcandy/Python_language_Answers/_0002.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        tmp = 0
        res = None
        now = None
        while (l1 is not None or l2 is not None or carry !=0):
            if l1 is not None and l2 is not None:
                tmp = l1.val + l2.val + carry
                l1 = l1.next
                l2 = l2.next
            elif l1 is None and l2 is None:
                tmp = carry
                if tmp == 0:
                    return res
            elif l1 is not None:
                tmp = l1.val + carry 
                l1 = l1.next
            elif l2 is not None:
                tmp = l2.val + carry
                l2 = l2.next
            print(tmp)
            if tmp >= 10:
                tmp -= 10
                carry = 1
            else:
                carry = 0
            if res is None:
                res = ListNode(tmp, None)
                now = res
            else:
                now.next = ListNode(tmp, None)
                now = now.next
        return res 
